### ISSUE TITLE
IIS challenge handler: Allow .well-known to be a virtual directory

### ISSUE THREAD

**Issue by rpresser-aleyant** [ebekker/ACMESharp#177]
My sites are handled in a load-balanced environment with two nodes, so I cannot guarantee that the node I'm running the Complete-ACMEChallenge script on is the one that will be seen when the Let's Encrypt server checks my submitted challenge. Looking at the code, it appears the IIS challenge handler looks up the filesystem folder for the root of the site, then creates a subfolder ".well-known" in that filesystem folder. I need it to be able to handle the case where ".well-known" is a virtual directory ... thus my two nodes can share the same ".well-known" folder and either one could handle the incoming HTTP request.

As an alternative I can manually robocopy the .well-known folder, but I think the virtual directory idea is more elegant. And more scalable if I should have more than two nodes.

**Comment by ebekker** [comment]
Can I ask, how is making the `.well-known` folder a virtual directory going to help you?  It still needs to *point* somewhere on disk where the challenge response will be created.  And on the second node, you'll still need to replicate the vdir configuration, correct?

**Comment by rpresser-aleyant** [comment]
1) It would point at a location on my file server. The virtual directory would point at the *same* folder from each node.
2) As my asp.net application goes through releases, the root folder of my site changes. But the virtual directory need not ever change.
3) Replicating virtual directories is part of my release process (thank you, octopusdeploy) and adding one more vdir to the process is no hardship.

BTW, today I used the robocopy technique and it worked just fine.

**Comment by ebekker** [comment]
One thing to consider, the IIS challenge handler already does support a `OverrideSiteRoot` provider parameter that allows you to override the path to the root of the web site referenced.  This could be a UNC path to a network share vs. a local path on a single server.

Once this is done, you only need to create the VDir on each server to point to that UNC path and you should be golden.  If you're using the POSH library for ACMESharp, then adding the VDir generation using POSH should be relatively simple as part of your overall script.

**Comment by rpresser-aleyant** [comment]
Hmm ... my nodes have site root on the local machine, I was thinking of just the vdir being on UNC. i.e.:

mysite.com/ => c:\octopus\applications\.....\version
mysite.com/.well-known => \\server\\path\.well-known

I'll look into it some more. Maybe staying with robocopy isn't such a bad idea ...

**Comment by ebekker** [comment]
No, that's all good -- the `OverrideSiteRoot` flag doesn't impact where you actually have your root.

See, normally the IIS provider will find your site by name or ID from the local IIS configuration, then use the root directory from that config to figure out where to write out the `.well-known` directory content.  But the `OverrideSiteRoot` simply changes where the IIS provider will resolve the path to write out to.

So basically this is what you were originally asking for, minus actually creating the VDir entry in the target IIS site, which you would have to do for all nodes in your farm, including the current one you're running ACMESharp from.

**Comment by rpresser-aleyant** [comment]
Ahhh! That makes great sense, thanks.

I never meant to be asking for it to create vdirs for me .. just that it follow the vdirs that are already created in IIS. This is just as good then.

**Comment by ebekker** [comment]
Cool!

**Comment by rpresser-aleyant** [comment]
This gave me a "Specified cast is not valid." exception:

`Complete-ACMEChallenge $domain -ChallengeType http-01 -Handler iis -HandlerParameters @{ WebSiteRef = "Aleyant.Pressero.Web.Store.production"; OverrideSiteRoot = "\\afs\p" }`

How should I be providing the handler parameter?
